src/Cle.SemanticAnalysis/SsaConverter.cs

```diff
@@ -0,0 +1,400 @@
+                var block = _originalMethod.Body.BasicBlocks[blockIndex];
+                Debug.Assert(block != null);
+
+                if (!_isSealed[blockIndex])
+                {
+                    // Incomplete CFG
+                    // Create a value for the Phi and set the Phi aside to be created when the block is sealed
+                    var phiType = _originalMethod.Values[variableIndex].Type;
+                    var phiValueNumber = CreateValueNumber(phiType, LocalFlags.None);
+                    WriteVariable(variableIndex, blockIndex, phiValueNumber);
+
+                    if (_incompletePhis[blockIndex] is null)
+                    {
+                        _incompletePhis[blockIndex] = new List<(ushort, ushort)>();
+                    }
+                    _incompletePhis[blockIndex].Add((variableIndex, phiValueNumber));
+
+                    return phiValueNumber;
+                }
+                else if (block.Predecessors.Count == 1)
+                {
+                    // The trivial case: no phi needed
+                    return ReadVariable(variableIndex, block.Predecessors[0]);
+                }
+                else
+                {
+                    // Complete CFG
+                    // Before recursing, write the Phi value to break cycles
+                    var phiType = _originalMethod.Values[variableIndex].Type;
+                    var phiValueNumber = CreateValueNumber(phiType, LocalFlags.None);
+                    WriteVariable(variableIndex, blockIndex, phiValueNumber);
+
+                    var operands = GetPhiOperands(variableIndex, blockIndex);
+                    if (operands.Count == 1)
+                    {
+                        // If there is exactly one operand, skip generating a Phi.
+                        // The value created earlier will be left unused.
+                        var trivialValue = (ushort)operands[0];
+                        WriteVariable(variableIndex, blockIndex, trivialValue);
+                        return trivialValue;
+                    }
+                    else
+                    {
+                        _blockGraphBuilder.GetBuilderByBlockIndex(blockIndex).AddPhi(phiValueNumber, operands);
+                        return phiValueNumber;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the list of operands for the given Phi, with duplicates removed.
+        /// </summary>
+        private ImmutableList<int> GetPhiOperands(ushort variableIndex, int blockIndex)
+        {
+            Debug.Assert(_originalMethod.Body != null);
+            var block = _originalMethod.Body.BasicBlocks[blockIndex];
+            Debug.Assert(block != null);
+
+            var operands = ImmutableList<int>.Empty.ToBuilder();
+            foreach (var predecessor in block.Predecessors)
+            {
+                // Skip duplicate operands
```

Comment:
I seem to have messed up a `git reset`, there should be a commit reverting this addition.